From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hayato Ito <hayato@chromium.org>
Date: Wed, 19 Aug 2026 22:40:58 -0700
Subject: [PATCH] [upstream] Handle synchronous completion in
 SetPlatformLocalNetworkAccessGranted

URLRequestHttpJob::SetPlatformLocalNetworkAccessGranted() used a
CHECK_EQ to verify that transaction_->RestartIgnoringLastError()
returned ERR_IO_PENDING. However, RestartIgnoringLastError() can return
error codes or complete synchronously. When it returns a value other
than ERR_IO_PENDING, the CHECK fails and causes a crash.

This change updates SetPlatformLocalNetworkAccessGranted() to handle
synchronous return values by posting a task to run OnStartCompleted,
matching the behavior in ContinueWithCertificate() and
ContinueDespiteLastError().

This fix was generated by the Crash Autofix pipeline for the crash
http://crash/12d3764df0400000. Please see go/crash-autofix for more
details.

Bug: 549297724
Change-Id: I616ddcb5dcd72dc818d888a388a59af6d16d508a
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8260580
Reviewed-by: Kenichi Ishibashi <bashi@chromium.org>
Commit-Queue: Hayato Ito <hayato@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1682876}
(cherry picked from commit 385fbe4301b45308c5f3530e9d7b91bcaca888bc)
---
 net/url_request/url_request_http_job.cc | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/net/url_request/url_request_http_job.cc b/net/url_request/url_request_http_job.cc
index e029de5a2abe7..9deaa7b28a5b4 100644
--- a/net/url_request/url_request_http_job.cc
+++ b/net/url_request/url_request_http_job.cc
@@ -1758,9 +1758,15 @@ void URLRequestHttpJob::SetPlatformLocalNetworkAccessGranted() {
 
   int rv = transaction_->RestartIgnoringLastError(base::BindOnce(
       &URLRequestHttpJob::OnStartCompleted, base::Unretained(this)));
-  // RestartIgnoringLastError() always returns ERR_IO_PENDING. See
-  // HttpNetworkTransaction.
-  CHECK_EQ(rv, ERR_IO_PENDING);
+  if (rv == ERR_IO_PENDING) {
+    return;
+  }
+
+  // The transaction started synchronously, but we need to notify the
+  // URLRequest delegate via the message loop.
+  TaskRunner(priority_)->PostTask(
+      FROM_HERE, base::BindOnce(&URLRequestHttpJob::OnStartCompleted,
+                                weak_factory_.GetWeakPtr(), rv));
 }
 
 void URLRequestHttpJob::CancelPlatformLocalNetworkAccessRequest() {
